### DISCUSSION TITLE
Dynamic Header/Footer Height

### DISCUSSION THREAD

**Discussion by procodix** [dompdf/dompdf#3477] [Show and tell]
I wanted to have fixed headers and footers on every page, without the need to manually apply a padding to the page. I mean, computers were invented to take care of calculations. I opened a few issues here regarding auto-height headers and footers, but they got closed or ignored. So here is a solution for others.

It was necessary to modify the reflow / render process in dompdf but I wanted to achieve it without patching the code. Finally got it working with just the provided events / callbacks.

The logic behind that is as follows: on page 0 only (!), the code walks through all HTML nodes and searches for header and footer elements with position:fixed assigned. Once found it simulates a rendering by calling the page's reflow and then reset. In between it captures the height of all header and footer elements and adds them to $headerheight or $footerheight. These are then applied as margin_top / margin_bottom to all pages, making the process automatic.

If you wanted different headers and footers on pages - you had to do the calculations on every page again before setting the margins. The idea wasn't tested but it could be working with some sort of header/footer:nth-child(x) {} styling for the different pages

```php
$pagecount = 0;
$headerheight = 0;
$footerheight = 0;
 $dompdf->setCallbacks([
            [
                'event' => 'begin_page_reflow',
                'f' => function($page) use (&$pagecount, &$headerheight, &$footerheight) {
					if ($pagecount == 0) {
						foreach ($page->get_children() as $f) {
							if (!in_array($f->get_node()->nodeName, ['header', 'footer'])) continue;
							if ($f->get_style()->position != 'fixed') continue;
							$f->reflow();
							$decorator = $f->get_frame()->get_decorator();
							if (!$decorator instanceof \Dompdf\FrameDecorator\Block) continue;
							foreach ($decorator->get_line_boxes() as $line) {
								if ($f->get_node()->nodeName == 'header') $headerheight += $line->h;
								if ($f->get_node()->nodeName == 'footer') $footerheight += $line->h;
							}
							$f->reset();
						}
					}
					$page->get_style()->margin_top = $headerheight;
					$page->get_style()->margin_bottom = $footerheight;
                },
            ],
            [
                'event' => 'end_document',
                'f' => function($pn, $pc, $cpdf) {
                    $canvas = $cpdf;
					$page = 0;
                    foreach ($canvas->get_cpdf()->objects as &$object) {
                        if ($object['t'] == 'page') $page++;
                        if ($object['t'] == 'contents') $object['c'] = str_replace([static::PAGECOUNT, static::PAGE], [$pc, $pn], $object['c']);
                    }
                },
            ]
        ]);
```

**Comment by bsweeney** [comment]
So the idea is that you don't know the size of the header ahead of time and this adjusts the page margins based on the calculated height of the header after manual reflow. Quite brilliant!

I have a sample on here somewhere about dynamically showing/hiding headers based on the current page. Something like this could be used to adjust the page margins based on the visible elements.

**Comment by procodix** [comment]
This could be integrated as default behavior for header or footer elements with height:auto

**Comment by comhon-project** [comment]
@procodix That exactly what i was looking for! thanks! 
I made an improvement to simplify and take margin, padding and border of the header and footer in account : 
(if I understand correctly, `get_margin_height` do the job)
```PHP
'f' => function ($page) use (&$pagecount, &$headerheight, &$footerheight) {
    if ($pagecount == 0) {
        $nodes = ['header', 'footer'];
        foreach ($page->get_children() as $frame) {
            if (!in_array($frame->get_node()->nodeName, $nodes) || $frame->get_style()->position != 'fixed') {
                continue;
            }
            $frame->reflow();
            if ($frame->get_node()->nodeName == 'header') {
                $headerheight = $frame->get_margin_height();
            }
            if ($frame->get_node()->nodeName == 'footer') {
                $footerheight = $frame->get_margin_height();
            }
            $frame->reset();
        }
    }
    $page->get_style()->margin_top = $headerheight;
    $page->get_style()->margin_bottom = $footerheight;
};
```

**Comment by comhon-project** [comment]
Starting from version 3.1.1, the proposed solutions no longer work, as fixed elements are no longer part of the list of children.
[commit](https://github.com/dompdf/dompdf/commit/ddff1c846ede0f1e689b3ce123abb0c8d5a09699)

**Reply by bsweeney** [reply]
The issue is that the `begin_page_reflow` event is called before the positioned elements are inserted into the current page. I think it makes sense to modify the logic flow to move that event below the element insertion.

follow #3665 for updates

**Reply by comhon-project** [reply]
fixed in version 3.1.3
